faunadb/client.go

```diff
@@ -235,6 +262,98 @@ func (client *FaunaClient) NewWithObserver(observer ObserverCallback) *FaunaClie
 	return client.newClient(client.basicAuth, observer)
 }
 
+// Stream creates a stream subscription to the result of the given expression.
+//
+// The subscription returned by this method does not issue any requests until
+// the subscription's Start method is called. Make sure to
+// subscribe to the events of interest, otherwise the received events are simply
+// ignored.
+func (client *FaunaClient) Stream(query Expr, config ...StreamConfig) StreamSubscription {
+	return newSubscription(client, query, config...)
+}
+
+
+func (client *FaunaClient) startStream(subscription *StreamSubscription) (err error) {
+	if subscription.Status() != StreamConnIdle {
+		err = errors.New("stream subscription already started")
+		return
+	}
+
+	startTime := time.Now()
+	if payload, err := client.prepareRequestBody(subscription.query); err == nil {
+		body := ioutil.NopCloser(bytes.NewReader(payload))
+
+		var endpoint strings.Builder
+		endpoint.WriteString(client.streamEndpoint)
+
+		if len(subscription.config.Fields) > 0 {
+			endpoint.WriteString("?fields=")
+			count := len(subscription.config.Fields)
+			for i := 0; i < count; i++ {
+				endpoint.WriteString(string(subscription.config.Fields[i]))
+				if i <= count-1 {
+					endpoint.WriteString(",")
+				}
+			}
+		}
+
+		subscription.status.Set(StreamConnOpening)
+		if response, err := client.performRequest(body, endpoint.String(), true, nil); err == nil {
+			httpResponse := response.response
+			_ = client.storeLastTxnTime(httpResponse.Header)
+			if err = checkForResponseErrors(httpResponse); err != nil {
+
+				subscription.status.Set(StreamConnError)
+
+				httpResponse.Body.Close()
+				response.cncl()
+				return err
+			}
+
+			go func() {
+				subscription.status.Set(StreamConnActive)
+
+				defer httpResponse.Body.Close()
+				defer response.cncl()
+
+				for {
```

Comment:
Q: What instruction in this for loop blocks until there's another message to be read?

Reply:
There is client.callObserver function in parseResponse.

Reply:
That's a driver specific callback isn't it? I don't think that's blocking this loop until next event arrives. There are two concerns here: (a) the spinning loop consuming CPU; b) your genera inability to stop this loop at command.

I don't think (a) is happening but I want us to have clarity on which part of it is blocking and waiting for next event to arrive. On (b) I think we need a close signal that we can monitor so that this coroutine can be closed when the subscription is closed.